doc/rst/meta/compiler-internals-doxygen/queries.rst

```diff
@@ -0,0 +1,288 @@
+ * returns a const reference to the desired result type
+ * is written in a particular way to interact with the Context through
+   the ``QUERY_`` macros defined in query-impl.h.
+
+Queries cannot be regular methods but they can be class methods (static
+methods).
+
+For example, here is a query that computes MyResultType from myArg1 and
+myArg2:
+
+.. code-block:: c++
+
+    #include "chpl/queries/query-impl.h"
+
+    const MyResultType& myQueryFunction(Context* context,
+                                        MyArgType MyArg1,
+                                        MyOtherArgType MyArg2) {
+      QUERY_BEGIN(myQueryFunction, context, myKey1, myKey2)
+
+      // do steps to compute the result
+      MyResultType result = ...;
+      // if an error is encountered, it can be saved with QUERY_ERROR(error)
+
+      return QUERY_END(result);
+    }
+
+To call the query, just write e.g. ``myQueryFunction(context, arg1, arg2)``.
+
+The macros are where the query framework takes important steps to make
+this function in to a query.
+
+In ``QUERY_BEGIN``, the query framework will check to see if there is
+already a result in the Context for these arguments that can be reused.
+If a result can be reused, ``QUERY_BEGIN`` will run the ``mark`` function
+on the result and return it. Otherwise, the query proceeds to execute the
+rest of the function body in order to compute the result.
+
+The framework also supports input queries which use ``QUERY_BEGIN_INPUT``
+instead of ``QUERY_BEGIN``. These queries pull in state from outside the
+framework - for example, they might read a file. They will be re-run
+every revision even if their dependencies are unchanged.
+
+While running the rest of the function body, any queries called will be
+automatically recorded as dependencies.
+
+The query concludes with ``return QUERY_END(result)`` which indicates to
+the query framework that ``result`` is a local variable containing the
+result of the query. This result, or one equivalent to it, will be
+returned and stored in the Context for future reuse. This process is
+managed by the ``update`` function.
+
+Types used as argument or return types in queries need to have certain
+functionality as described in the Requirements on Types Used in Queries
+section below.
+
+Checking and Recomputing Queries
+--------------------------------
+
+The first step in a query occurs within ``QUERY_BEGIN``. In this step,
+the query framework checks to see if the body of the query needs to be
+run.  This uses ``lastChecked`` and ``lastChanged`` values which are
+revision numbers stored for the query in the Context.
+
+The framework will check if dependencies have changed and this process
+will include recursively re-running them if necessary.
+
+The framework will skip running the body and return an existing
+result if:
+
+ * lastChecked matches the current revision number
+ * the query is not an input query and no dependencies changed in this
+   revision
```

Comment:
maybe add how lastChanged fits in this. Maybe something like "eg. lastChanged < thisRevision"